### PR TITLE
align with node streams

### PR DESCRIPTION
**Pull request by ronag**
This refactors destroy to better align with the node stream invariants.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -96,7 +101,6 @@ class MultiStream extends stream.Readable {
   _gotNextStream (stream) {
     if (!stream) {
       this.push(null)
-      this.destroy()
```
Comment (ronag) [inline index.js]: Use autoDestroy to ensure proper event ordering

File: index.js [hunk 2]
```diff
@@ -108,8 +112,13 @@ class MultiStream extends stream.Readable {
...
     const onClose = () => {
-      if (!stream._readableState.ended) {
-        this.destroy()
+      const readableEnded =
+        (stream._readableState && stream._readableState.ended) ||
+        stream.readableEnded
+      if (!readableEnded && !stream.destroyed) {
+        const err = new Error('ERR_STREAM_PREMATURE_CLOSE')
+        err.code = 'ERR_STREAM_PREMATURE_CLOSE'
+        this.destroy(err)
```
Comment (ronag) [inline index.js]: close without end is an error

File: index.js [hunk 3]
```diff
@@ -118,6 +127,7 @@ class MultiStream extends stream.Readable {
       stream.removeListener('readable', onReadable)
       stream.removeListener('end', onEnd)
       stream.removeListener('close', onClose)
+      stream.destroy()
```
Comment (ronag) [inline index.js]: not all streams will auto destroy on 'end', so we should explicitly destroy just to make sure we don't have any dangling resources in buggy streams

File: index.js [hunk 4]
```diff
@@ -58,19 +57,25 @@ class MultiStream extends stream.Readable {
     this._forwarding = false
   }
 
-  destroy (err) {
```
Comment (ronag) [inline index.js]: We should not override `destroy` as that will bypass important teardown logic and event ordering.

File: index.js [hunk 5]
```diff
@@ -58,19 +57,25 @@ class MultiStream extends stream.Readable {
...
+    if (typeof this._queue !== 'function') streams = streams.concat(this._queue)
+
+    if (streams.length === 0) {
+      cb(err)
+    } else {
+      let counter = streams.length
+      streams.forEach(stream => {
+        destroy(stream, err, err => {
+          if (err || --counter === 0) {
+            cb(err)
```
Comment (mafintosh) [inline index.js]: this can call the cb more than once if multiple streams errors

Comment (ronag) [reply index.js]: See `cb = once(cb)` at the top of the function

File: index.js [hunk 6]
```diff
@@ -143,3 +153,16 @@ MultiStream.obj = streams => (
 )
 
 module.exports = MultiStream
+
+// Normalize stream destroy w/ callback.
+function destroy (stream, err, cb) {
```
Comment (mafintosh) [inline index.js]: if the stream is already destroyed you need to return early i think

File: index.js (hunk 5 above)
Comment (mafintosh) [reply index.js]: 👍 in either case it should not call the cb until all streams have called back tho

File: index.js (hunk 6 above)
Comment (ronag) [reply index.js]: not for native `destroy`, however, overriden `destroy` might not behave as expected so it's probably a good idea

File: index.js (hunk 5 above)
Comment (ronag) [reply index.js]: good point

Comment (ronag) [reply index.js]: fixed

File: index.js (hunk 6 above)
Comment (ronag) [reply index.js]: fixed

File: package.json [hunk 7]
```diff
@@ -7,6 +7,7 @@
     "url": "https://github.com/feross/multistream/issues"
   },
   "dependencies": {
+    "once": "^1.4.0",
```
Comment (mafintosh) [inline package.json]: still used?

Comment (ronag) [reply package.json]: destroy wrapper, line 162

File: index.js [hunk 8]
```diff
@@ -108,8 +112,13 @@ class MultiStream extends stream.Readable {
     }
 
     const onClose = () => {
-      if (!stream._readableState.ended) {
-        this.destroy()
+      const readableEnded =
+        (stream._readableState && stream._readableState.ended) ||
+        stream.readableEnded
```
Comment (mafintosh) [inline index.js]: _readableState is always set as all streams are run through the toStreams2 thing above. i don't see why the readableEnded check is needed but can't hurt i guess :)

**Comment by mafintosh** [comment]
I think this looks good. 

I'll let @feross click review as I don't want to imply any warranty 🤣 I get enough "frustrated stream breakage github issues" as is.

**Comment by ronag** [comment]
@mafintosh I took this one step further... though I'm not sure if it's maybe too much changes... what do you think https://github.com/ronag/multistream/commit/900ac9d78baffa22201e8cdf57602771c5d46fb2?

**Comment by feross** [comment]
FYI - I won't have time to review this for a while. Will do as soon as I can.

**Comment by pcothenet** [comment]
@feross @ronag this fixes an important issue for us. Not sure how tbh, but a recent update introduced a "premature close" error. Using @ronag branch fixes it.

**Comment by alxhotel** [comment]
I agree with @pcothenet.

This PR also fixes the `premature close` error in Webtorrent.

**Comment by ronag** [comment]
I'm more than happy to help out here. But repo admin needs to review and merge.

**Comment by goto-bus-stop** [comment]
@feross will you have time for this in the near future? the PR seems to work pretty well on my end!

**Comment by artemirq** [comment]
I'd recommend to use this approach, for me it works in more cases correctly.

```js
const merge = (...streams) => {
  let pass = new PassThrough();
  let waiting = streams.length;

  for (const stream of streams) {
    pass = stream.pipe(pass, { end: false });
    stream.once('end', () => --waiting === 0 && pass.emit('end'));
  }

  return pass;
};
```

Given from https://stackoverflow.com/questions/16431163/concatenate-two-or-n-streams

**Comment by feross** [comment]
I'm happy to merge this, but the conflicts need to be fixed.

@mafintosh Did this look good to merge to you?

File: index.js [hunk 9]
```diff
@@ -143,3 +150,16 @@ MultiStream.obj = streams => (
 )
 
 module.exports = MultiStream
+
+// Normalize stream destroy w/ callback.
+function destroy (stream, err, cb) {
+  if (!stream.destroy || stream.destroyed) {
+    cb(err)
+  } else {
+    const callback = once(er => cb(er || err))
```
Comment (mafintosh) [inline index.js:159]: this wont trip up with the boolean argument from socket.on('close') ? https://nodejs.org/dist/latest-v14.x/docs/api/all.html#net_event_close_1

Comment (ronag) [reply index.js:159]: it will

File: index.js [hunk 10]
```diff
@@ -143,3 +150,16 @@ MultiStream.obj = streams => (
...
+    stream
+      .on('error', callback)
+      .on('close', callback)
```
Comment (ronag) [inline index.js:162]:
```suggestion
      .on('close', () => callback()))
```

**Comment by mafintosh** [comment]
👍 from me

**Comment by feross** [comment]
@ronag Looks like this still has conflicts – can you resolve those and then this looks ready to merge!

File: index.js [hunk 11]
```diff
@@ -144,3 +151,16 @@ MultiStream.obj = streams => (
...
+// Normalize stream destroy w/ callback.
+function destroy (stream, err, cb) {
+  if (!stream.destroy || stream.destroyed) {
+    cb(err)
+  } else {
+    const callback = once(er => cb(er || err))
+    stream
+      .on('error', callback)
+      .on('close', () => callback()))
+      .destroy(err, callback)
```
Comment (feross) [inline index.js:164]: Does the `destroy()` method really take a `callback`?

File: index.js [hunk 12]
```diff
@@ -144,3 +151,16 @@ MultiStream.obj = streams => (
...
+
+// Normalize stream destroy w/ callback.
+function destroy (stream, err, cb) {
+  if (!stream.destroy || stream.destroyed) {
+    cb(err)
+  } else {
+    const callback = once(er => cb(er || err))
+    stream
+      .on('error', callback)
+      .on('close', () => callback()))
```
Comment (feross) [inline index.js:163]: You have an extra closing paren here causing a syntax error.

**Review by feross — CHANGES_REQUESTED** [review]
Looks good – just one final bit of feedback

**Comment by feross** [comment]
Sorry the delay everyone. This is now merged and released as 4.1.0!

File: index.js (hunk 12 above)
Comment (feross) [reply index.js:163]: I fixed it

**Pull request merged**